CAIP-25: Simplified and pragmatic spec

pedrouid opened the pull request:
TLDR'
- Reduced verbosity of most sections to make it easier to read
- Renamed:
    - `optionalScopes` to `scopes`
    - `scopedProperties` to `capabilities`
    - `sessionProperties` to `properties`
- Removed `requiredScopes` and retained only `scopes` (fka `optionalScopes`).
- Merged `capabilities` (fka `scopedProperties`) into `scopeObjects`.
- Removed Namespace-scoped `scopeObjects` and retained only Chain-scoped `scopeObjects`
- Remove chain prefix for accounts inside `scopeObjects`

jiexi commented:
thoughts on dropping the session prefix? `sessionScopes` -> `scopes`, `sessionCapabilities` -> `capabilities`?

I'm still unsure if i like the rename from properties to capabilities. Will have to think more.

While we're making changes, what are your thoughts on removing the `references` shorthand? Sure, it kind of makes sense convenience wise for humans manually forming these requests, but I don't think that's a scenario we should optimize for. A machine can easily form the correct request object without the `references` shorthand. Removing the `references` shorthand simplifies the validation and handling logic for the implementors of CAIP-25

adonesky1 commented on CAIPs/caip-25.md:
```diff
@@ -16,110 +16,90 @@ CAIP-25 defines an authorization procedure for a chain agnostic provider to inte
 ## Motivation
 
-The motivation comes from the lack of standardization across blockchains to expose accounts and define the expected JSON-RPC methods to be used by an application through a provider connecting to a signer or other user agent.
+The lack of standardization across blockchains for exposing accounts and defining expected JSON-RPC methods has created inconsistencies. CAIP-25 addresses this by enabling a unified, session-based interface between applications and wallets.
```
Maybe?
```suggestion
The absence of standardized interfaces and abstractions for reading from and writing to blockchains—such as consistent account models and JSON-RPC method specifications—has fragmented application and wallet interactions. CAIP-25 resolves this by defining a unified, session-based interface that standardizes communication between applications and wallets.
```

adonesky1 commented on CAIPs/caip-25.md:
```diff
@@ -16,110 +16,90 @@ CAIP-25 defines an authorization procedure for a chain agnostic provider to inte
-    * [`wallet_getSession`][CAIP-312] to enable the caller to query for the current status of the session at any time.
-    * [`wallet_revokeSession`][CAIP-285] to explicitly end the session
-    * [`wallet_sessionChanged`][CAIP-311] to notify caller of updated session authorizations.
+The session is proposed by a caller and accepted by the respondent. The respondent may return a `sessionId` which both parties then persist along with session properties and authorization scopes.
```
should we atleast reference [CAIP-316](https://chainagnostic.org/CAIPs/caip-316)
```suggestion
The session is proposed by a caller and accepted by the respondent. The respondent may return a `sessionId` which both parties then persist along with session properties and authorization scopes. See [CAIP-316][] for guidance on session lifecycles with and without `sessionId`s.
```

adonesky1 commented on CAIPs/caip-25.md:
```diff
@@ -130,54 +110,17 @@ Example:
-Each object should be keyed to the scope of a `sessionScopes` member to which it corresponds.
-All properties of each object in `scopedProperties` MUST be interpreted by the respondent as proposals or declarations rather than as requirements.
-In addition to making additional properties of or metadata about the corresponding `sessionScopes` member explicit, they can also annotate, support, or extend the negotiation of scope proposals (e.g., providing connection information about unfamiliar scopes, or which accounts to expose to each).
+The `sessionCapabilities` object MAY be present and provides metadata or annotations tied to specific `sessionScopes` keys.
```
Seems like we're removing some possibly important language around the optionality of these capabilities? And perhaps should be more specific about what kinds of annotations/metadata we would expect to be put here?

pedrouid commented:
Actually I think its a good idea to drop the references shorthand and just make it all flat into single-scope scopes

If that is the case then we can also include capabilities inside the scope and even rename it to chains

pedrouid commented:
Thanks again @jiexi @adonesky1 for the first round of comments!

1. Updated the "Motivation" was suggested in the comments
2. Included a reference to CAIP-316 for session lifecyscles
3. Remove Namespace-scoped in favor of Chain-scoped scopes
4. Removed `session` prefix from top-level fields
5. Merged `capabilities` into `scopedObjects`
6. Included both @jiexi and @adonesky1 as authors

jiexi commented on CAIPs/caip-25.md:
```diff
@@ -16,422 +16,247 @@ CAIP-25 defines an authorization procedure for a chain agnostic provider to inte
-
-The successful result MUST contain an object called `sessionScopes` which contains 1 or more `scopeObjects`.
-- All required `scopeObjects` and all, none, or some of the optional `scopeObject`s (at the discretion of the provider) MUST be included if successful.
-- Unlike the request, each scope object MUST also contain an `accounts` array,
```
is this being loosened?

jiexi commented on CAIPs/caip-25.md:
```diff
@@ -16,422 +16,247 @@ CAIP-25 defines an authorization procedure for a chain agnostic provider to inte
-The respondent SHOULD ignore and drop from its response any properties not defined in this document or in another CAIP document extending this protocol which the respondent has implemented in its entirety;
-similarly, the `requiredScopes`, `optionalScopes`, and `sessionScopes` objects returned by the respondent SHOULD contain only valid [CAIP-217][] objects, and properties not defined in [CAIP-217][] SHOULD also be dropped from each of those objects.
-The same absolute security posture is not expected for the metadata objects `scopedProperties` and `sessionProperties`, but caution is still recommended for such extensions:
-callers and respondents alike SHOULD allow for their counterparties dropping or ignoring unfamiliar members from either.
```
i feel like some of this guidance is helpful. Is it worth giving up these sections to make the spec appear simpler?

bumblefudge replied:
Move them to non-normative appendices, maybe?

jiexi commented on CAIPs/caip-25.md:
```diff
@@ -16,422 +16,247 @@ CAIP-25 defines an authorization procedure for a chain agnostic provider to inte
+Each `scopeObject` is keyed by a [CAIP-2][] chain ID.
 
-The distinction between `requiredScopes` and `optionalScopes` is ultimately semantic, since a wallet may still choose to establish a connection authorizing a subset of requested networks or requested capabilities from each; the primary function of the distinction is to offer callers a mechanism for signaling which scopes they consider primary and which they consider secondary to their request, in ...
+Wallets MAY authorize a subset of scopes. This enables granular control and flexibility.
```
We allow subset, superset, or entirely disjoint set to be granted. The dapp requested scopes are only taken as a suggestion on what defaults to show to the user. The user is free to deviate from the default as they wish.

jiexi replied:
This wording works for me

jiexi commented on CAIPs/caip-25.md:
```diff
@@ -16,422 +16,247 @@ CAIP-25 defines an authorization procedure for a chain agnostic provider to inte
-`scopeObjects` keyed to an entire [CAIP-104][] namespace SHOULD contain a non-empty `references` array to be actionable, making them functionally equivalent to a series of identical `scopeObjects`, each keyed to one of the members of `references` expressed as a [CAIP-2][] scope.
-An empty or absent `references` array SHOULD NOT be interpreted as a namespace-wide authorization (i.e. authorization for ANY network therein), but rather as a null authorization of 0 specified `references`s within that namespace.
-(See [CAIP-217][] for more details on the structure of the typed objects included in these `...Scopes` objects.)
+Each `scopeObject` is keyed by a [CAIP-2][] chain ID.
```
This touches on how we don't have a good place to park methods that do not belong to any ecosystem/chain. Right now we're using the `wallet` scope for that, but it is sort of a pain as it isn't a CAIP-2 chain ID.

The other issue is that we don't have a good place to park methods/accounts that belong to an ecosystem but to no specific chain. In the example below, I see `eip155:0` used, but AFAICT the `0` reference is only meant to be used when specifying that an CAIP-10 Account ID is an EOA account. I feel that this is implied both by the wording in the [EIP-155 CAIP-10 spec](https://github.com/ChainAgnostic/namespaces/blob/main/eip155/caip10.md), but also that this specifically is specified on the CAIP-10 spec and not the CAIP-2 spec

bumblefudge replied:
Good point.  I deliberately only defined the "CAIP-10" usage of chainId 0 at the time because it was only discussed in the context of EOAs (and the older Gnosis Safe convention).  Adding a similar section to the EIP-155 CAIP-2 spec would work if CAIP-2 wanted to be able to use `eip155:0` specifically as a key for `scopeObject`s, for example.  If that's cleaner or more durable long-term than keeping the "wallet:eip155" special-case, let's discuss.

jiexi commented:
adding this here before I forget. Should we revisit if `accounts` needs to be full CAIP-10` addresses since they are nested in a CAIP-2 scope? I can see an argument for and against.

bumblefudge commented on CAIPs/caip-25.md:
```diff
@@ -16,422 +16,247 @@ CAIP-25 defines an authorization procedure for a chain agnostic provider to inte
 
-This proposal has the goal to define a standard procedure for decentralized applications to interface with chain agnostic cryptocurrency wallets and other user agents which govern identities (including accounts) in multiple cryptographic systems.
-It defines a lightweight protocol for negotiating and persisting authorizations during a session managed by a provider construct.
+This proposal defines a standard procedure for decentralized applications to interface with chain agnostic cryptocurrency wallets and other user agents that govern identities (including accounts) across multiple cryptographic systems. It specifies a lightweight protocol for negotiating and persisting authorizations during a session managed by a provider construct.
```
```suggestion
This proposal defines a standard procedure for decentralized applications to interface with chain agnostic cryptocurrency wallets and other user agents that govern identities (including accounts) across multiple cryptographic systems. It specifies a lightweight protocol for negotiating and persisting authorizations during a session managed either by an in-DOM provider construct, by a securely-addressed browser extensions, or by a distinct user-agent.
```

bumblefudge commented on CAIPs/caip-25.md:
```diff
@@ -16,422 +16,247 @@ CAIP-25 defines an authorization procedure for a chain agnostic provider to inte
-When a caller wishes revoke an active session, it can do so by calling [`wallet_revokeSession`][CAIP-285].
-- When a `sessionId` is returned in the initial `wallet_createSession` response, the caller MUST call `wallet_revokeSession` with the supplied `sessionId` to revoke that session.
-- When the wallet does not provide a `sessionId` in its initial response, a call to `wallet_revokeSession` revokes the single active session between caller and wallet.
+Session updates triggered by the wallet must notify the caller using `wallet_sessionChanged`.
```
```suggestion
Session updates initiated by the wallet must notify the caller using `wallet_sessionChanged`.
```

bumblefudge commented on CAIPs/caip-25.md:
```diff
@@ -16,422 +16,247 @@ CAIP-25 defines an authorization procedure for a chain agnostic provider to inte
+Session updates triggered by the wallet must notify the caller using `wallet_sessionChanged`.
 
-For more detail on the lifecycle and management of sessions with and without `sessionId`s, see the informational [CAIP-316][].
+Callers may revoke sessions using `wallet_revokeSession`, either by `sessionId` or directly if no sessionId was established.
```
```suggestion
Callers may revoke sessions using `wallet_revokeSession`, passing the `sessionId` parameter if it was returned by the initial response.
```

bumblefudge commented on CAIPs/caip-25.md:
```diff
@@ -16,422 +16,247 @@ CAIP-25 defines an authorization procedure for a chain agnostic provider to inte
+        "methods": ["wallet_grantPermissions"],
+        "notifications": [],
+        "capabilities": {
+          "signInWithEthereum": {
```
could go either way, or even allow both if it's written down anywhere obvious that "signInWithEthereum" is an alias for "eip4361".  i see the benefit of aligning with the Coinbase-wallet/5792 capability names and/or aliasing the more educational spec-rooted names to them.

jiexi replied:
i'm also fine with not using specs as the name of the field, but i'd prefer clearer guidance from us on how to extend this in the future? Do we require an information CAIP for each capability?

pedrouid commented:
> adding this here before I forget. Should we revisit if `accounts` needs to be full CAIP-10` addresses since they are nested in a CAIP-2 scope? I can see an argument for and against.

I've updated the example response to reflect this

adonesky1 commented on CAIPs/caip-25.md:
````diff
@@ -108,6 +112,8 @@ If a connection is rejected, the wallet MAY respond with a generic error or sile
 }
 ```
 
+The `wallet` object MUST be included for wallet-level method and notification requests.
````
I'm not totally clear what this means. Must it be included if there are no `wallet-level` methods or notifications?

jiexi commented on CAIPs/caip-25.md:
```diff
@@ -16,422 +16,229 @@ CAIP-25 defines an authorization procedure for a chain agnostic provider to inte
-          "foo": "bar"
+    "properties": {
+      "expiry": "2022-12-24T17:07:31+00:00",
+      "walletInfo": {
+        "uuid": "350670db-19fa-4704-a166-e52e178b59d2",
+        "name": "Example Wallet",
+        "icon": "data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg'/>",
+        "rdns": "com.example.wallet"
```
i like this over here a lot more than in the `wallet` scope

pedrouid replied:
glad to hear that

adonesky1 commented on CAIPs/caip-25.md:
```diff
@@ -16,422 +16,229 @@ CAIP-25 defines an authorization procedure for a chain agnostic provider to inte
+        "capabilities": {}
+      },
+      "wallet:eip155": {
+        "methods": ["wallet_grantPermissions", "wallet_getAssets"],
+        "notifications": [],
+        "capabilities": {
+          "walletService": "https://wallet-service.example.com/rpc"
+        }
+      },
+      "eip155:1": {
+        "accounts": ["0xab16a96d359ec26a11e2c2b3d8f8b8942d5bfcdb"],
+        "methods": ["eth_sendTransaction", "personal_sign"],
         "notifications": ["accountsChanged", "chainChanged"],
-        "accounts": ["eip155:1:0xab16a96d359ec26a11e2c2b3d8f8b8942d5bfcdb", "eip155:137:0xab16a96d359ec26a11e2c2b3d8f8b8942d5bfcdb"]
+        "capabilities": {}
```
I left this comment somewhere before, but `accounts` should be required in each scopeObject here... with the possible exception of the new `wallet:any` scope

adonesky1 replied:
oh wait nevermind I see they are here now, just moved... I'm just being dense. Do we need to specify somewhere how/why the `wallet:*` scopes don't need accounts though?

bumblefudge replied:
couldn't they be a required array but empty array is a valid value? that feels more explicit, like "I AM GIVING YOU ZERO ACCOUNTS FOR THIS" 😅 

(and, sidenote, might a wallet namespace be the right place to stick an ephemeral/implicit/offchain-only address used for completely off-node usecases like presenting verifiable credentials? a girl can dream...)

adonesky1 commented on CAIPs/caip-25.md:
```diff
@@ -16,422 +16,229 @@ CAIP-25 defines an authorization procedure for a chain agnostic provider to inte
+Each entry within `scopes` object MAY contain `accounts` and `capabilities` as part of its object for success response.
 
-##### Generic Failure Code
+#### Error
```
?
```suggestion
#### Error Codes
```

pedrouid commented on CAIPs/caip-25.md:
```diff
@@ -16,422 +16,229 @@ CAIP-25 defines an authorization procedure for a chain agnostic provider to inte
-        "methods": ["eth_sendTransaction", "eth_signTransaction", "eth_sign", "get_balance", "personal_sign"],
-        "notifications": ["accountsChanged", "chainChanged"]
+    "scopes": {
+      "wallet:any": {
```
Agreed

bumblefudge replied:
`eip155:any` would also be good to include in examples
